controllers/provider-aws/pkg/controller/backupbucket/actuator.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backupbucket
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+
+	"github.com/gardener/gardener-extensions/controllers/provider-aws/pkg/aws"
+	awsclient "github.com/gardener/gardener-extensions/controllers/provider-aws/pkg/aws/client"
+	"github.com/gardener/gardener-extensions/pkg/controller/backupbucket"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+type actuator struct {
+	backupbucket.Actuator
+	client client.Client
+	logger logr.Logger
+}
+
+func newActuator() backupbucket.Actuator {
+	return &actuator{
+		logger: log.Log.WithName("aws-backupbucket-actuator"),
+	}
+}
+
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+func (a *actuator) Reconcile(ctx context.Context, bb *extensionsv1alpha1.BackupBucket) error {
+	providerSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, kutil.Key(bb.Spec.SecretRef.Namespace, bb.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	awsClient, err := awsclient.NewClient(string(providerSecret.Data[aws.AccessKeyID]), string(providerSecret.Data[aws.SecretAccessKey]), bb.Spec.Region)
+	if err != nil {
+		return err
+	}
+
+	return awsClient.CreateBucket(ctx, bb.Name)
+}
+
+func (a *actuator) Delete(ctx context.Context, bb *extensionsv1alpha1.BackupBucket) error {
+	providerSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, kutil.Key(bb.Spec.SecretRef.Namespace, bb.Spec.SecretRef.Name), providerSecret); err != nil {
+		return err
+	}
+
+	awsClient, err := awsclient.NewClient(string(providerSecret.Data[aws.AccessKeyID]), string(providerSecret.Data[aws.SecretAccessKey]), bb.Spec.Region)
+	if err != nil {
+		return err
+	}
+
+	return awsClient.DeleteBucket(ctx, bb.Name)
```

Comment:
Before deleting the bucket, shouldn't you check that no `BackupEntry` objects reference this bucket anymore?

Reply:
I plan to have that check only for BackupBucket resource in Garden cluster i.e. as a part of BackupBucket controller of GCM and not gardener-extension.

Reply:
Hm, not sure if that's a good idea to only do it that way. IMHO you need both checks to be safer.

Reply:
I haven't added it here. Cause backupEntry might be in different seed other than backup bucket. And hence backup bucket won't have access to all backupEntries. The check in GCM for the same handles that safely.

Reply:
Thanks, makes sense, agreed.